02_Architecture/02_Getting_Some_Output.md

```diff
@@ -0,0 +1,113 @@
+# Serial Logging (aka Hello World)
+
+During the developmen of our kernel we will need to debug a lot, and checking a lot of values, but so far our kernel is not capable of doing anything, and having proper video output with scrolling, fonts etc, can take some time to implement (and probably we need to check again some values), so we need a quick way of getting some output from our kernel, not necessarily on the screen. 
+
+This is where the serial logging came to an aid, we will use the serial port to output our text and numbers. 
+
+Many emulators has an option to redirect serial output to a file, if you are using QEmu (for more information about it refer to the Appendices section) you need to start it passing the parameter *-s filename*:
+
+```bash
+qemu -S filename.log -cdrom yourosiso
+```
+
+This will save the serial output on the file called `filename.log`, if we want the serial output directly on the screen, we can use `stdio` instead.
+
+## Printing to Serial
+
+We will use the `inb` and `outb` instruction to communicate with the serial port. But the first thing our kernel should do is do is being able to write to serial ports. To do that we need: 
+
+* for simiplicity and readability two C functions that will make use of the inb/outb asm instructions (luckily they are asm functions so making their c version is very easy)
+* initialization of serial communication
+* and at least an instruction to send characters and strings to the serial. 
+
+The first step is pretty strightforward, using inline assembly we will create two "one-line" functions for inb and outb: 
+
+```C
+extern inline unsigned char inportb (int portnum)
+{
+  unsigned char data=0;
+  __asm__ __volatile__ ("inb %%dx, %%al" : "=a" (data) : "d" (portnum));
+  return data;
+}
+
+extern inline void outportb (int portnum, unsigned char data)
+{
+  __asm__ __volatile__ ("outb %%al, %%dx" :: "a" (data),"d" (portnum));
+}
+
+```
+
+### Initialization
+
+The second part again is pretty simple, we just need to send few configuration command for initializing the serial communication, the code below is copied from https://wiki.osdev.org/Serial_Ports#Initialization:
```

Comment:
I would mention debugcon here too, since it requires no setup (less room for error) and is often faster than the emulated serial chip. To use it simply write to port 0xE9.